src/passes/Monomorphize.cpp

```diff
@@ -47,44 +81,218 @@
+      // value sent from the call.
+      operands.push_back(ExpressionManipulator::flexibleCopy(
+        operand, wasm, [&](Expression* child) -> Expression* {
+          if (canBeMovedIntoContext(child)) {
+            // This can be moved, great: let the copy happen.
+            return nullptr;
+          }
+
+          // This cannot be moved, so we stop here: this is a value that is sent
+          // into the monomorphized function. It is a new operand in the call,
+          // and in the context operands it is a local.get, that reads that
+          // value.
+          auto paramIndex = newOperands.size();
+          newOperands.push_back(child);
+          // TODO: If one operand is a tee and another a get, we could actually
+          //       reuse the local, effectively showing the monomorphized
+          //       function that the values are the same. (But then the checks
+          //       later down to is<LocalGet> would need to check index too.)
+          return builder.makeLocalGet(paramIndex, child->type);
+        }));
+    }
+
+    // TODO: handle drop
+    dropped = false;
+  }
+
+  // Checks whether an expression can be moved into the context.
+  bool canBeMovedIntoContext(Expression* curr) {
+    // Constant numbers, funcs, strings, etc. can all be copied, so it is ok to
+    // add them to the context.
+    return Properties::isSingleConstantExpression(curr);
+  }
+
+  // Check if a context is trivial relative to a call, that is, the context
+  // contains no information that can allow optimization at all. Such trivial
+  // contexts can be dismissed early.
+  bool isTrivial(Call* call, Module& wasm) {
+    // Dropped contexts are not trivial.
+    if (dropped) {
+      return false;
+    }
+
+    // The context must match the call for us to compare them.
+    assert(operands.size() == call->operands.size());
+
+    // If an operand is not simply passed through, then we are not trivial.
+    auto callParams = wasm.getFunction(call->target)->getParams();
+    for (Index i = 0; i < operands.size(); i++) {
+      // A local.get of the same type implies we just pass through the value.
+      // Anything else is not trivial.
+      if (!operands[i]->is<LocalGet>() || operands[i]->type != callParams[i]) {
+        return false;
+      }
+    }
+
+    // We found nothing interesting, so this is trivial.
+    return true;
+  }
+};
+
+} // anonymous namespace
+
+} // namespace wasm
+
+namespace std {
+
+template<> struct hash<wasm::CallContext> {
+  size_t operator()(const wasm::CallContext& info) const {
+    size_t digest = hash<bool>{}(info.dropped);
+
+    wasm::rehash(digest, info.operands.size());
+    for (auto* operand : info.operands) {
+      wasm::hash_combine(digest, wasm::ExpressionAnalyzer::hash(operand));
+    }
+
+    return digest;
+  }
+};
+
+// Useful for debugging.
+[[maybe_unused]] std::ostream& operator<<(std::ostream& o,
+                                          wasm::CallContext& context) {
```

Comment:
If you name this `dump()` or something, it will be callable from a debugger, and therefore even more useful.

Reply:
I never use a debugger myself, so I would never have thought of that :smile: Done.